src/ol_infrastructure/substructure/keycloak/__main__.py

```diff
@@ -145,3 +148,140 @@
     login_theme="keycloak",
     opts=resource_options,
 )
+
+# Create OL Public Realm
+ol_apps_realm = keycloak.Realm(
+    "olapps",
+    access_code_lifespan="30m",
+    access_code_lifespan_user_action="15m",
+    attributes={
+        "business_unit": f"operations-{env_name}",
+    },
+    display_name="OLApps",
+    display_name_html="<b>MIT Open Learning Apps</b>",
+    enabled=True,
+    login_theme="base",
+    duplicate_emails_allowed=False,
+    otp_policy=keycloak.RealmOtpPolicyArgs(
+        algorithm="HmacSHA256",
+        digits=6,
+        initial_counter=2,
+        look_ahead_window=1,
+        period=30,
+        type="totp",
+    ),
+    realm="olapps",
+    reset_password_allowed=True,
+    verify_email=True,
+    password_policy="upperCase(2) and digits(2) and length(12) and specialChars(2) and forceExpiredPasswordChange(365) and notUsername and notEmail",  # noqa: E501,S106 # pragma: allowlist secret
+    security_defenses=keycloak.RealmSecurityDefensesArgs(
+        brute_force_detection=keycloak.RealmSecurityDefensesBruteForceDetectionArgs(
+            failure_reset_time_seconds=43200,
+            max_failure_wait_seconds=900,
+            max_login_failures=10,
+            minimum_quick_login_wait_seconds=60,
+            permanent_lockout=True,
+            quick_login_check_milli_seconds=1000,
+            wait_increment_seconds=60,
+        ),
+        headers=keycloak.RealmSecurityDefensesHeadersArgs(
+            content_security_policy="frame-src 'self' https://www.google.com/recaptcha/; frame-ancestors 'self'; object-src 'none'; script-src 'self' https://www.google.com/recaptcha/;",  # noqa: E501
+            content_security_policy_report_only="",
+            strict_transport_security="max-age=31536000; includeSubDomains",
+            x_content_type_options="nosniff",
+            x_frame_options="https://www.google.com",  # CAPTCHA
+            x_robots_tag="none",
+            x_xss_protection="1; mode=block",
+        ),
+    ),
+    smtp_server=keycloak.RealmSmtpServerArgs(
+        auth=keycloak.RealmSmtpServerAuthArgs(
+            password=email_password,
+            username=email_username,
+        ),
+        from_="odl-devops@mit.edu",
+        from_display_name="Identity - Open Learning Platform Engineering",
+        host=email_host,
+        port=587,
+        reply_to="odl-devops@mit.edu",
+        reply_to_display_name="Identity - Open Learning Platform Engineering",
+        starttls=True,
+    ),
+    ssl_required="external",
+    offline_session_idle_timeout="168h",
+    sso_session_idle_timeout="2h",
+    sso_session_max_lifespan="24h",
+    opts=resource_options,
+)
+
+required_action_configure_otp = keycloak.RequiredAction(
```

Comment:
Is this setting MFA as a required step for users? I'm not sure that we want to force that on everyone. I think it's a good idea, but we might want to give folks an option of not configuring it.

Reply:
Does that allow for the use of TOTP but not require it?

Reply:
So new users won't be forced to enable TOTP, but they do have the option to go to their account settings and add an authenticator app.

Reply:
Sounds good, thanks :+1: